plugins/c8y_firmware_plugin/src/download.rs

```diff
@@ -0,0 +1,336 @@
+use c8y_api::smartrest::smartrest_serializer::SmartRestSerializer;
+use c8y_api::smartrest::smartrest_serializer::SmartRestSetOperationToExecuting;
+use c8y_api::smartrest::smartrest_serializer::SmartRestSetOperationToFailed;
+use c8y_api::smartrest::smartrest_serializer::SmartRestSetOperationToSuccessful;
+use c8y_api::smartrest::smartrest_serializer::TryIntoOperationStatusMessage;
+use mqtt_channel::Message;
+use mqtt_channel::SinkExt;
+use mqtt_channel::Topic;
+use mqtt_channel::UnboundedSender;
+use sha256::digest;
+use sha256::try_digest;
+use std::collections::HashMap;
+use std::fs;
+use std::os::unix::fs as unix_fs;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use tedge_api::OperationStatus;
+use tedge_utils::timers::Timers;
+use tokio::sync::Mutex;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+#[cfg(not(test))]
+use tedge_config::DEFAULT_FILE_TRANSFER_ROOT_PATH;
+
+#[cfg(not(test))]
+pub const FILE_TRANSFER_ROOT_PATH: &str = DEFAULT_FILE_TRANSFER_ROOT_PATH;
+#[cfg(not(test))]
+pub const FILE_TRANSFER_CACHE_PATH: &str = "/var/tedge/cache";
+#[cfg(test)]
+pub const FILE_TRANSFER_ROOT_PATH: &str = "/tmp/file-transfer";
+#[cfg(test)]
+pub const FILE_TRANSFER_CACHE_PATH: &str = "/tmp/cache";
+
+pub struct FirmwareDownloadManager {
+    tedge_device_id: String,
+    mqtt_publisher: UnboundedSender<Message>,
+    http_client: Arc<Mutex<dyn C8YHttpProxy>>,
+    local_http_host: String,
+    tmp_dir: PathBuf,
+    pub operation_timer: Timers<(String, String), ActiveOperationState>,
+    pub url_map: HashMap<String, String>,
+    pub timeout_sec: Duration,
+}
+
+impl FirmwareDownloadManager {
+    pub fn new(
+        tedge_device_id: String,
+        mqtt_publisher: UnboundedSender<Message>,
+        http_client: Arc<Mutex<dyn C8YHttpProxy>>,
+        local_http_host: String,
+        tmp_dir: PathBuf,
+        timeout_sec: Duration,
+    ) -> Self {
+        FirmwareDownloadManager {
+            tedge_device_id,
+            mqtt_publisher,
+            http_client,
+            local_http_host,
+            tmp_dir,
+            operation_timer: Timers::new(),
+            url_map: HashMap::new(),
+            timeout_sec,
+        }
+    }
+
+    pub async fn handle_firmware_download_request(
+        &mut self,
+        smartrest_request: SmartRestFirmwareRequest,
+    ) -> Result<(), anyhow::Error> {
+        info!(
+            "Handling c8y_Firmware operation: device={}, name={}, version={}, url={}",
+            smartrest_request.device,
+            smartrest_request.name,
+            smartrest_request.version,
+            smartrest_request.url,
+        );
+
+        if smartrest_request.device == self.tedge_device_id {
+            warn!("c8y-firmware-plugin does not support firmware operation for the main tedge device. \
+            Please define a custom operation handler for the c8y_Firmware operation.");
+            Ok(())
```

Comment:
Seeing that `Ok(())` here surprised me a bit as I thought that it'd make the operation succeed in the cloud. But since that's not the case, it's fine. May be worth adding a comment that we're ignoring this request so that some other plugin on the machine can handle it.